pkg/network/protocols/http/shared_libraries_test.go

```diff
@@ -160,25 +165,252 @@ func TestSameInodeRegression(t *testing.T) {
+			checkPIDNotAssociatedWithPathID(watcher, fooPathID1, uint32(command1.Process.Pid)) ||
+			checkPIDNotAssociatedWithPathID(watcher, fooPathID2, uint32(command1.Process.Pid)) {
+			return false
+		}
+
+		return int64(1) == registers.Load()
+	}, time.Second*10, time.Second, "")
+	require.Len(t, watcher.registry.byID, 1)
+	require.NoError(t, command1.Process.Kill())
+
+	require.Eventuallyf(t, func() bool {
+		// Checking path1 still exists, and path2 not.
+		return checkPathIDDoesNotExist(watcher, fooPathID1) && checkPathIDDoesNotExist(watcher, fooPathID2) &&
+			checkPIDNotAssociatedWithPathID(watcher, fooPathID1, uint32(command1.Process.Pid)) &&
+			checkPIDNotAssociatedWithPathID(watcher, fooPathID2, uint32(command1.Process.Pid))
+	}, time.Second*10, time.Second, "")
+}
+
+func TestSoWatcherLeaks(t *testing.T) {
+	perfHandler := initEBPFProgram(t)
+
+	fooPath1, fooPathID1 := createTempTestFile(t, "foo.so")
+	fooPath2, fooPathID2 := createTempTestFile(t, "foo2.so")
+
+	registerCB := func(id pathIdentifier, root string, path string) error { return nil }
+	unregisterCB := func(id pathIdentifier) error { return nil }
+
+	watcher := newSOWatcher(perfHandler,
+		soRule{
+			re:           regexp.MustCompile(`foo.so`),
+			registerCB:   registerCB,
+			unregisterCB: unregisterCB,
+		},
+		soRule{
+			re:           regexp.MustCompile(`foo2.so`),
+			registerCB:   registerCB,
+			unregisterCB: unregisterCB,
+		},
+	)
+	watcher.Start()
+
+	// create files
+	clientBin := buildSOWatcherClientBin(t)
+
+	command1 := exec.Command(clientBin, fooPath1, fooPath2)
+	require.NoError(t, command1.Start())
+
+	// Check sowatcher map
+	require.Eventuallyf(t, func() bool {
+		// Checking both paths exist.
+		if checkPathIDDoesNotExist(watcher, fooPathID1) || checkPathIDDoesNotExist(watcher, fooPathID2) {
+			return false
+		}
+
+		// Checking the PID associated with the 2 paths.
+		return checkPIDAssociatedWithPathID(watcher, fooPathID1, uint32(command1.Process.Pid)) &&
+			checkPIDAssociatedWithPathID(watcher, fooPathID2, uint32(command1.Process.Pid))
+	}, time.Second*10, time.Second, "")
+
+	command2 := exec.Command(clientBin, fooPath1)
+	require.NoError(t, command2.Start())
+
+	require.Eventuallyf(t, func() bool {
+		// Checking both paths exist.
+		if checkPathIDDoesNotExist(watcher, fooPathID1) || checkPathIDDoesNotExist(watcher, fooPathID2) {
+			return false
+		}
+
+		// Checking PID1 is still associated to the 2 paths, and PID2 is associated only with the first path
+		return checkPIDAssociatedWithPathID(watcher, fooPathID1, uint32(command1.Process.Pid)) &&
+			checkPIDAssociatedWithPathID(watcher, fooPathID2, uint32(command1.Process.Pid)) &&
+			checkPIDAssociatedWithPathID(watcher, fooPathID1, uint32(command2.Process.Pid)) &&
+			checkPIDNotAssociatedWithPathID(watcher, fooPathID2, uint32(command2.Process.Pid))
+	}, time.Second*10, time.Second, "")
+
+	require.NoError(t, command1.Process.Kill())
```

Comment:
This could leak processes if the previous `require` statements evaluate to false. I suggest including a `t.Cleanup` call to kill the started processes.